vllm/distributed/kv_transfer/kv_connector/v1/nixl_connector.py

```diff
@@ -412,75 +425,94 @@ def __init__(self, vllm_config: VllmConfig, engine_id: str):
-        metadata = handshake(path, 0)
+        # TODO: make the scheme dynamic, and/or implement https on both sides.
+        url = build_uri("http", host, port, path="get_kv_connector_metadata")
+        logger.debug("Querying metadata on path: %s", url)
+
+        try:
+            req = Request(url)
+            with urlopen(req, timeout=5.0) as response:
+                response_data = response.read().decode('utf-8')
+                res = json.loads(response_data)
+                logger.debug("NIXL handshake response: %s", res)
+        except (URLError, HTTPError) as e:
+            logger.error("Failed to fetch metadata from %s: %s", url, e)
+            raise
+
+
+        remote_tp_size = len(res.keys())
+        # Default case is that the remote TP size is 1, so we can
+        # directly access the metadata.
+        tp_data = res.get(str(self.tp_rank), {}).get("0", {})
+        metadata_bytes = tp_data.get("agent_metadata", None)
 
         # Handshake only with the other TP remote the current local rank will
         # pull from. With homogeneous TP it happens to be the same rank_i.
-        tp_ratio = self._tp_size[self.engine_id] // metadata.tp_size
+        tp_ratio = self._tp_size[self.engine_id] // remote_tp_size
         p_remote_rank = self.tp_rank // tp_ratio
         if p_remote_rank > 0:
-            path = make_zmq_path("tcp", host, port + p_remote_rank)
-            logger.debug("Querying metadata on path: %s at remote rank %s",
-                         path, p_remote_rank)
-            _ = handshake(path, p_remote_rank)
+            metadata_bytes = res.get(str(p_remote_rank), {}).get("0", {}).get(
+                "agent_metadata", None)
+
+        if metadata_bytes is not None:
+            # Reconstruct NixlAgentMetadata from JSON response
+            # agent_metadata is base64-encoded binary data, not msgpack
+            metadata = NixlAgentMetadata(
+                engine_id=tp_data["engine_id"],
+                agent_metadata=base64.b64decode(metadata_bytes),
+                kv_caches_base_addr=tp_data["kv_caches_base_addr"],
+                num_blocks=tp_data["num_blocks"],
+                tp_size=tp_data["tp_size"],
+                block_len=tp_data["block_len"],
+                attn_backend_name=tp_data["attn_backend_name"],
+            )
+
+            # Register Remote agent.
+            logger.debug("About to register remote agent for engine %s", 
+                        metadata.engine_id)
+            pre_register = time.perf_counter()
+            self.add_remote_agent(metadata, remote_tp_rank=p_remote_rank)
+            agent_time = time.perf_counter()
+            logger.debug("Finished registering remote agent for engine %s", 
+                        metadata.engine_id)
+
+            logger.debug("NIXL handshake: get metadata took: %s", 
+                        pre_register - start_time)
+            logger.debug("NIXL handshake: add agent took: %s", 
```

Comment:
just curious, is this value changing significantly with http?

Reply:
It's slightly slower w/ nixl 0.2, with nixl 0.3.1 agent registration is a lot faster, which makes the increase more noticable. But, since we are moving this onto a background thread, so the engine is not blocked during handshake. If we can reduce the amount of metadata needed for handshake (like the optimization you already have) I think we can get closer to zmq speed.